File: test/cpu/addi.cc
#define BOOST_TEST_DYN_LINK
#include "68040.hpp"

#include "test.hpp"
#include <boost/test/data/monomorphic.hpp>
#include <boost/test/data/test_case.hpp>
#include <boost/test/unit_test.hpp>
namespace bdata = boost::unit_test::data;

struct F_ADDI {
    F_ADDI() {
        // ADDI.B #100, %D2
        TEST::SET_W(0, 0003000 | 2);
        TEST::SET_W(2, 100);
        TEST::SET_W(4, TEST_BREAK);

        // ADDI.W #30000, %D2
        TEST::SET_W(6, 0003100 | 2);
        TEST::SET_W(8, 30000);
        TEST::SET_W(10, TEST_BREAK);

        // ADDI.W #0x70000000, %D2
        TEST::SET_W(12, 0003200 | 2);
        TEST::SET_L(14, 0x70000000);
        TEST::SET_W(18, TEST_BREAK);

        jit_compile(0, 20);
    }
};
BOOST_FIXTURE_TEST_SUITE(ADDI, Prepare, *boost::unit_test::fixture<F_ADDI>())

BOOST_AUTO_TEST_SUITE(Byte)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 20;
    run_test(0);
    BOOST_TEST(cpu.D[2] == 120);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
    BOOST_TEST(!cpu.V);
}
BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = -100;
    run_test(0);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -110;
    run_test(0);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(CX) {
    cpu.D[2] = -92;
    run_test(0);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 60;
    run_test(0);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Word)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 2000;
    run_test(6);
    BOOST_TEST(cpu.D[2] == 32000);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
    BOOST_TEST(!cpu.V);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = -30000;
    run_test(6);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -32000;
    run_test(6);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(CX) {
    cpu.D[2] = -12000;
    run_test(6);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 10000;
    run_test(6);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE(Long)

BOOST_AUTO_TEST_CASE(value) {
    cpu.D[2] = 0x05000000;
    run_test(12);
    BOOST_TEST(cpu.D[2] == 0x75000000);
    BOOST_TEST(!cpu.Z);
    BOOST_TEST(!cpu.N);
    BOOST_TEST(!cpu.C);
    BOOST_TEST(!cpu.X);
    BOOST_TEST(!cpu.V);
}

BOOST_AUTO_TEST_CASE(Z) {
    cpu.D[2] = -0x70000000;
    run_test(12);
    BOOST_TEST(cpu.Z);
}
BOOST_AUTO_TEST_CASE(N) {
    cpu.D[2] = -0x72000000;
    run_test(12);
    BOOST_TEST(cpu.N);
}
BOOST_AUTO_TEST_CASE(CX) {
    cpu.D[2] = -0x32000000;
    run_test(12);
    BOOST_TEST(cpu.C);
    BOOST_TEST(cpu.X);
}
BOOST_AUTO_TEST_CASE(V) {
    cpu.D[2] = 0x30000000;
    run_test(12);
    BOOST_TEST(cpu.V);
}
BOOST_AUTO_TEST_SUITE_END()
BOOST_AUTO_TEST_SUITE_END()

File: test/lua/cpu/addi.lua
-- ADDI.B #10, %D1
MEM[0].W = OCT "003000" | 1
MEM[2].W = 10
MEM[4].W = TEST.BREAK

-- ADDI.W #30000, %D2
MEM[6].W = OCT "003100" | 2
MEM[8].H = 30000
MEM[10].W = TEST.BREAK

-- ADDI.L #1000000000, %D3
MEM[12].W = OCT "003200" | 3
MEM[14].I = 1000000000
MEM[18].W = TEST.BREAK

JIT_COMPILE(0, 20)

local run = {}

run.B = {
    value = function()
        CPU.D[2].C = 50
        TEST.RUN(0)
        TEST.CHECK("CPU.D[2].C", 60)
    end,
    z = farray(TF, function(b)
        CPU.D[2].C = cond(b, -10, 5)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[2].C = cond(b, -20, 10)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.N", b)
    end),
    cx = farray(TF, function(b)
        CPU.D[2].C = cond(b, -1, 20)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.C", b)
        TEST.CHECK("CPU.CC.X", b)
    end),
    v = farray(TF, function(b)
        CPU.D[2].C = cond(b, 120, 20)
        TEST.RUN(0)
        TEST.CHECK("CPU.CC.V", b)
    end)
}

run.W = {
    value = function()
        CPU.D[3].H = 2000
        TEST.RUN(6)
        TEST.CHECK("CPU.D[3].H", 32000)
    end,
    z = farray(TF, function(b)
        CPU.D[3].H = cond(b, -30000, 5)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[3].H = cond(b, -31000, 10)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.N", b)
    end),
    cx = farray(TF, function(b)
        CPU.D[3].H = cond(b, -1, 2000)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.C", b)
        TEST.CHECK("CPU.CC.X", b)
    end),
    v = farray(TF, function(b)
        CPU.D[3].H = cond(b, 3000, 20)
        TEST.RUN(6)
        TEST.CHECK("CPU.CC.V", b)
    end)
}

run.L = {
    value = function()
        CPU.D[4].I = 500000000
        TEST.RUN(12)
        TEST.CHECK("CPU.D[4].I", 1500000000)
    end,
    z = farray(TF, function(b)
        CPU.D[4].I = cond(b, -1000000000, 5)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.Z", b)
    end),
    n = farray(TF, function(b)
        CPU.D[4].I = cond(b, -2000000000, 10)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.N", b)
    end),
    cx = farray(TF, function(b)
        CPU.D[4].I = cond(b, -1, 3)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.C", b)
        TEST.CHECK("CPU.CC.X", b)
    end),
    v = farray(TF, function(b)
        CPU.D[4].I = cond(b, 0x7FFFFFFF, 20)
        TEST.RUN(12)
        TEST.CHECK("CPU.CC.V", b)
    end)
}
return run
